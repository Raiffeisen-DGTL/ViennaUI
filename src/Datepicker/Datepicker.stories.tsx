import React, { useRef } from 'react';
import { Story, Meta } from 'storybook';
import { Datepicker, DatepickerProps } from './Datepicker';
import { Button } from '@/Button';

export default {
    title: 'Development/Datepicker',
    component: Datepicker,
} as Meta;

export const Overview: Story<DatepickerProps> = (args) => {
    const [value, setValue] = React.useState();
    const ref = useRef<any>();

    const handleChange = React.useCallback(({ value: data }) => setValue(data), []);
    return <Datepicker ref={ref} value={value} onChange={handleChange} />;
};

export const ViewOnly: Story<DatepickerProps> = (args) => {
    return <Datepicker viewOnly value={new Date('2020-10-10')} />;
};

export const Year5000Bug: Story<DatepickerProps> = () => {
    const [date, setDate] = React.useState<any>();

    return (
        <Datepicker
            autoComplete='off'
            min={undefined}
            max={undefined}
            value={date}
            onChange={({ value, event, options }) => {
                setDate(options.date);
            }}
        />
    );
};

export const TestDatepicker = () => {
    const initialDate = new Date();
    const [date, setDate] = React.useState<Date | undefined>(initialDate);
    const handleDatepickerChange: DatepickerProps['onChange'] = ({ options: { date } }) => {
        console.log('date change', date);
        setDate(date);
    };

    return (
        <div>
            <Datepicker value={date} onChange={handleDatepickerChange} />
        </div>
    );
};

export const DatepickerBugWithViewOnly: Story<DatepickerProps> = (args) => {
    const [value, setValue] = React.useState();
    const [viewOnly, setViewOnly] = React.useState(false);
    const handleChange = React.useCallback(({ value }) => setValue(value), []);
    return (
        <>
            <Button onClick={() => setViewOnly((p) => !p)} size='m'>
                Change viewOnly
            </Button>
            <Datepicker value={value} onChange={handleChange} viewOnly={viewOnly} />
        </>
    );
};
