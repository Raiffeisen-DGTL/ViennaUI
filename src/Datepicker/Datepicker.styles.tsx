import styled, { css } from 'styled-components';
import { datepicker } from 'vienna.ui-theme';
import { getPresets } from '../Utils/styling';

const presets = getPresets(
    datepicker,
    'datepicker'
)({
    calendar: null,
});

export const Box = styled.div`
    position: relative;
    display: flex;
    flex-direction: column;
    outline: none;
`;

export const InputBox = styled.div`
    position: relative;
`;

export interface CalendarBoxProps {
    $hidden: boolean;
}
export const CalendarBox = styled.div<CalendarBoxProps>`
    ${presets.calendar};
    ${({ $hidden }) =>
        $hidden &&
        css`
            display: none;
        `};
`;
