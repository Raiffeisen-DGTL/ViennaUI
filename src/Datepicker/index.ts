export { Datepicker, type DatepickerProps } from './Datepicker';
export { type DatePickerLocalization } from './localization';
