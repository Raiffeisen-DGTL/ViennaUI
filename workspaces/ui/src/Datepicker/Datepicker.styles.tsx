import styled from 'styled-components';
import { getPresets } from 'vienna.ui-primitives';

const presets = getPresets('datepicker', {
    calendar: null,
});

export const Box = styled.div`
    position: relative;
    display: flex;
    flex-direction: column;
    outline: none;
`;

export const InputBox = styled.div`
    position: relative;
`;

export const CalendarBox = styled.div`
    position: absolute;
    opacity: 0;
    overflow: hidden;
    ${presets.calendar};
`;
