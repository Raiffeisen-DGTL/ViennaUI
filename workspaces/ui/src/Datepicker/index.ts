export * from './Datepicker';
export type { DatePickerLocalization } from './localization';
