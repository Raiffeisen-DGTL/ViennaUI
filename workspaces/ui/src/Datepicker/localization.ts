import { DSLocalizationProps } from '../Localization';
import { CalendarLocalization } from '../Calendar';
import { InputDateLocalization } from '../InputMask';

export type DatePickerLocalization = CalendarLocalization & InputDateLocalization;

export type DatePickerLocalizationProps = DSLocalizationProps<DatePickerLocalization>;
