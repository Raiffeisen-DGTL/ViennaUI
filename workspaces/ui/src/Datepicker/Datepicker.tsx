import React, { useCallback, useEffect, useMemo, useRef, useState } from 'react';
import { useDrop } from 'vienna.react-use';
import { Calendar as CalendarIcon } from 'vienna.icons';
import { Locale } from 'date-fns';
import { DatePickerLocalizationProps } from './localization';
import { Calendar, dateFunction, DateValue, DisabledDates, eventDateFunction } from '../Calendar';
import { InputDate } from '../InputMask';
import { getDateFromString, getStringFromDate } from '../Utils/DateUtils/DateUtils';
import { InputEvent, InputProps } from '../Input';
import { Box, InputBox, CalendarBox } from './Datepicker.styles';
import { checkIsDisabled } from '../Calendar/Utils';

export interface DatepickerProps
    extends Omit<InputProps, 'value' | 'type' | 'onChange' | 'onPaste'>,
        DatePickerLocalizationProps {
    /**
     * дата передается строкой вида 01.01.2021
     */
    value?: string | Date;

    /**
     * Отображение календаря
     */
    isCalendarOpen?: boolean;

    /**
     * Даты-событие
     */
    eventDates?: Date[] | eventDateFunction;

    /**
     * Выходные даты
     */
    weekendDates?: DisabledDates.WEEKENDS | Date[] | dateFunction;

    /**
     * Отображение кнопки "Сегодня"
     */
    todayButton?: boolean;

    /**
     * Нижняя граница выбора даты
     */
    minDate?: Date;

    /**
     * Верхняя граница выбора даты
     */
    maxDate?: Date;

    onChange?: (
        event: InputEvent<React.FormEvent<HTMLInputElement>> | Event | null,
        data: { value?: string; name?: string; isDisabled?: boolean; date?: Date }
    ) => void;

    /**
     * Неактивные для выбора даты
     */
    disabledDates?: DisabledDates.WEEKENDS | Date[] | dateFunction;

    /**
     * Дни недели отображаются с понедельника (1) или с воскресенья (0)
     */
    startingWeekDay?: 0 | 1;
    /**
     * Локаль календаря
     */
    locale?: Locale;
}

export const Datepicker: React.FC<DatepickerProps> = ({
    eventDates,
    isCalendarOpen,
    maxDate,
    minDate,
    name,
    postfix,
    prefix,
    size,
    todayButton,
    value,
    onChange,
    onFocus,
    onBlur,
    onKeyPress,
    weekendDates,
    disabledDates,
    tabIndex,
    startingWeekDay = 1,
    localization,
    locale,
    ...attrs
}: DatepickerProps): JSX.Element => {
    const [isOpen, setOpen] = useState<boolean | undefined>(isCalendarOpen);
    const [isDisabled, setIsDisabled] = useState<boolean>(false);
    const [active, setActive] = useState(false);
    const datepickerEl = useRef<HTMLDivElement>(null);
    const calendarRef = useDrop({ align: 'vertical', float: 'start', margins: { x: 0, y: 4 } });

    const handleInputFocus = useCallback(
        (event, data) => {
            if (typeof onFocus === 'function') {
                onFocus(event, { value: data.value, name });
            }
            setOpen(true);
        },
        [name, onFocus]
    );

    const handleBlur = useCallback(
        (event) => {
            if (typeof onBlur === 'function') {
                onBlur(event, { value: typeof value === 'string' ? value : getStringFromDate(value as Date), name });
            }
            setOpen(false);
            setActive(false);
        },
        [onBlur, value, name]
    );

    useEffect(() => {
        if (isOpen) {
            setActive(true);
        }
    }, [isOpen]);

    useEffect(() => {
        setOpen(isCalendarOpen);
    }, [isCalendarOpen]);

    const handleClickInput = useCallback(() => {
        if (active) {
            setOpen(true);
        }
    }, [isOpen, active]);

    const handleClickDateIcon = useCallback(() => {
        setOpen(!isOpen);
    }, [isOpen]);

    const handleChangeInput = useCallback(
        (event, data) => {
            const date = getDateFromString(data.value);
            const nextIsDisabled =
                !!date && checkIsDisabled({ dates: disabledDates, date, startingWeekDay, minDate, maxDate });
            if (typeof onChange === 'function') {
                onChange(event, { value: data.value, name, isDisabled: nextIsDisabled, date });
            }

            setIsDisabled(nextIsDisabled);
        },
        [name, onChange]
    );

    const handleClickDate = useCallback(
        (event, { date }: { date?: DateValue | Date }) => {
            if (date) {
                const clickedDateString = getStringFromDate(date as Date);

                if (typeof onChange === 'function') {
                    onChange(event, { value: clickedDateString, name, date: date as Date });
                }
                setOpen(false);
                setActive(true);
            }
        },
        [name, onChange]
    );

    const handleKeyPress = useCallback(
        (event) => {
            if (typeof onKeyPress === 'function') {
                onKeyPress(event);
            }
            if (event.key === 'Enter') {
                setOpen(!isOpen);
            }
        },
        [isOpen, onKeyPress]
    );

    const handleClickDocument = (event): void => {
        if (
            event.target instanceof Node &&
            datepickerEl &&
            datepickerEl.current &&
            !datepickerEl.current.contains(event.target)
        ) {
            setOpen(false);
            setActive(false);
        }
    };

    useEffect(() => {
        if (document.body) {
            document.body.addEventListener('click', handleClickDocument);
        }

        return () => {
            if (document.body) {
                document.body.removeEventListener('click', handleClickDocument);
            }
        };
    });

    const dateValue = useMemo(() => getDateFromString(value as any), [value, isDisabled]);

    const handleMouseDownCalendar = useCallback((e) => {
        e.preventDefault();
    }, []);

    return (
        <Box ref={datepickerEl} tabIndex={tabIndex} onBlur={handleBlur}>
            <InputBox onClick={handleClickInput}>
                <InputDate
                    postfix={
                        postfix ?? <CalendarIcon size={size === 'xxl' ? 'xl' : size} onClick={handleClickDateIcon} />
                    }
                    prefix={prefix}
                    size={size}
                    name={name}
                    value={value}
                    active={active}
                    localization={localization}
                    onChange={handleChangeInput}
                    onKeyPress={handleKeyPress}
                    onFocus={handleInputFocus}
                    {...attrs}
                />
            </InputBox>
            {isOpen && (
                <CalendarBox ref={calendarRef} onMouseDown={handleMouseDownCalendar}>
                    <Calendar
                        date={dateValue}
                        eventDates={eventDates}
                        weekendDates={weekendDates}
                        maxDate={maxDate}
                        minDate={minDate}
                        todayButton={todayButton}
                        disabledDates={disabledDates}
                        startingWeekDay={startingWeekDay}
                        localization={localization}
                        locale={locale}
                        onChange={handleClickDate}
                    />
                </CalendarBox>
            )}
        </Box>
    );
};

Datepicker.displayName = 'Datepicker';
Datepicker.defaultProps = {
    isCalendarOpen: false,
    todayButton: true,
};

export default Datepicker;
