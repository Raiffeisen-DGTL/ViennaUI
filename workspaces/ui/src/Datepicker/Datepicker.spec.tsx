/* eslint-disable @typescript-eslint/no-misused-promises */
import React from 'react';
import { Datepicker } from './Datepicker';

test('Datepicker', async () => {
    const snap = snapshot.render(<Datepicker />);
    expect(snap).toMatchSnapshot();
});

test('Datepicker /w not complete', async () => {
    const snap = snapshot.render(<Datepicker value='22.02' />);
    expect(snap).toMatchSnapshot();
});

test('Datepicker /w complete', async () => {
    const snap = snapshot.render(<Datepicker value='11.12.1900' />);
    expect(snap).toMatchSnapshot();
});

test('Datepicker /w date min and max valid', () => {
    const fn = function (e, data) {
        expect(data.value).toBe('15.06.2020');
    };
    const snap = snapshot.mount(
        <Datepicker minDate={new Date(2020, 5, 2)} maxDate={new Date(2020, 5, 15)} onChange={fn} />
    );

    const input = snap.find('input');
    input.simulate('change', { target: { value: '15.06.2020' } });
});

test('Datepicker /w date min and max invalid', () => {
    const fn = function (e, data) {
        expect(data.value).toBe('15.06.2020');
    };
    const snap = snapshot.mount(
        <Datepicker minDate={new Date(2020, 5, 2)} maxDate={new Date(2020, 5, 15)} onChange={fn} />
    );

    const input = snap.find('input');
    input.simulate('change', { target: { value: '15.06.2021' } });
});

test('Datepicker /w date not in range', async () => {
    const snap = await snapshot.mount(
        <Datepicker value={new Date(2020, 5, 17)} minDate={new Date(2020, 5, 2)} maxDate={new Date(2020, 5, 15)} />
    );
    expect(snap).toMatchSnapshot();
});

test('Datepicker /w date not in range and not fix value', async () => {
    const snap = await snapshot.mount(
        <Datepicker value={new Date(2020, 5, 17)} minDate={new Date(2020, 5, 2)} maxDate={new Date(2020, 5, 15)} />
    );
    expect(snap).toMatchSnapshot();
});

test('Datepicker /w invalid', async () => {
    const snap = snapshot.render(<Datepicker value={undefined} invalid />);
    expect(snap).toMatchSnapshot();
});
