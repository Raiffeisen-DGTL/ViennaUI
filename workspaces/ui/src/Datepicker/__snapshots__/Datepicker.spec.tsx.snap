// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Datepicker /w complete 1`] = `
<div
  class="Datepickerstyles__Box-sc-15d4dji-0 dajHNb"
>
  <div
    class="Datepickerstyles__InputBox-sc-15d4dji-1 jpzdav"
  >
    <label
      class="Inputstyles__Wrapper-hc8ttv-6 daAZxl"
    >
      <div
        class="Inputstyles__Wrapped-hc8ttv-5 lhoBGY"
      >
        <div
          class="Inputstyles__Box-hc8ttv-1 bzkyqR"
        >
          <div
            class="Inputstyles__PlaceholderWrapper-hc8ttv-7 frQCmf"
          >
            <div
              class="Inputstyles__Placeholder-hc8ttv-0 dDQzQZ"
            >
              11.12.1900
            </div>
          </div>
          <input
            aria-invalid="false"
            class="Inputstyles__StyledInput-hc8ttv-2 bUBKOu"
            value="11.12.1900"
          />
        </div>
      </div>
      <div
        class="Inputstyles__Postfix-hc8ttv-4 dvGfwK"
      >
        <svg
          cursor="pointer"
          fill="currentColor"
          height="20"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs />
          <path
            d="M10 3v2h4V3h2v2h4a1 1 0 011 1v14a1 1 0 01-1 1H4a1 1 0 01-1-1V6a1 1 0 011-1h4V3h2zm8.999 10h-14L5 19h14l-.001-6zM16 15a1 1 0 110 2 1 1 0 010-2zM8 7H5l-.001 4h14L19 7h-3v2h-2V7h-4v2H8V7z"
          />
        </svg>
      </div>
    </label>
  </div>
</div>
`;

exports[`Datepicker /w date not in range 1`] = `
<Datepicker
  isCalendarOpen={false}
  maxDate={2020-06-14T21:00:00.000Z}
  minDate={2020-06-01T21:00:00.000Z}
  todayButton={true}
  value={2020-06-16T21:00:00.000Z}
>
  <Datepickerstyles__Box
    onBlur={[Function]}
  >
    <div
      className="Datepickerstyles__Box-sc-15d4dji-0 dajHNb"
      onBlur={[Function]}
    >
      <Datepickerstyles__InputBox
        onClick={[Function]}
      >
        <div
          className="Datepickerstyles__InputBox-sc-15d4dji-1 jpzdav"
          onClick={[Function]}
        >
          <InputDate
            active={false}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyPress={[Function]}
            postfix={
              <Calendar
                color="currentColor"
                cursor="pointer"
                onClick={[Function]}
                size="m"
              />
            }
            value={2020-06-16T21:00:00.000Z}
          >
            <InputMask
              active={false}
              blocks={
                Object {
                  "DATE": Object {
                    "mask": [Function],
                    "max": 9998-12-31T21:00:00.000Z,
                    "min": 1899-12-31T21:29:43.000Z,
                  },
                }
              }
              design="outline"
              format={[Function]}
              lazy={true}
              mask="DATE"
              onChange={[Function]}
              onFocus={[Function]}
              onKeyPress={[Function]}
              parse={[Function]}
              postfix={
                <Calendar
                  color="currentColor"
                  cursor="pointer"
                  onClick={[Function]}
                  size="m"
                />
              }
              size="l"
              smartPlaceholder="17.06.2020"
              value="17.06.2020"
            >
              <Input
                active={false}
                design="outline"
                onChange={[Function]}
                onFocus={[Function]}
                onKeyPress={[Function]}
                onUpdated={[Function]}
                postfix={
                  <Calendar
                    color="currentColor"
                    cursor="pointer"
                    onClick={[Function]}
                    size="m"
                  />
                }
                size="l"
                smartPlaceholder="17.06.2020"
                value="17.06.2020"
              >
                <InputWrapper
                  active={false}
                  design="outline"
                  size="l"
                >
                  <Inputstyles__Wrapper
                    active={false}
                    design="outline"
                    size="l"
                  >
                    <label
                      className="Inputstyles__Wrapper-hc8ttv-6 daAZxl"
                      size="l"
                    >
                      <Inputstyles__Wrapped>
                        <div
                          className="Inputstyles__Wrapped-hc8ttv-5 lhoBGY"
                        >
                          <NativeInput
                            active={false}
                            aria-invalid={false}
                            design="outline"
                            key=".1"
                            onBlur={[Function]}
                            onChange={[Function]}
                            onFocus={[Function]}
                            onKeyPress={[Function]}
                            onUpdated={[Function]}
                            size="l"
                            smartPlaceholder="17.06.2020"
                            value="17.06.2020"
                          >
                            <Inputstyles__Box>
                              <div
                                className="Inputstyles__Box-hc8ttv-1 bzkyqR"
                              >
                                <Inputstyles__PlaceholderWrapper>
                                  <div
                                    className="Inputstyles__PlaceholderWrapper-hc8ttv-7 frQCmf"
                                  >
                                    <Inputstyles__Placeholder
                                      design="outline"
                                      size="l"
                                    >
                                      <div
                                        className="Inputstyles__Placeholder-hc8ttv-0 dDQzQZ"
                                        size="l"
                                      >
                                        <Inputstyles__Invisible>
                                          <span
                                            className="Inputstyles__Invisible-hc8ttv-8 bCdWvB"
                                          >
                                            17.06.2020
                                          </span>
                                        </Inputstyles__Invisible>
                                        <span />
                                      </div>
                                    </Inputstyles__Placeholder>
                                  </div>
                                </Inputstyles__PlaceholderWrapper>
                                <Inputstyles__StyledInput
                                  active={false}
                                  aria-invalid={false}
                                  design="outline"
                                  onBlur={[Function]}
                                  onChange={[Function]}
                                  onFocus={[Function]}
                                  onKeyPress={[Function]}
                                  onScroll={[Function]}
                                  size="l"
                                  value="17.06.2020"
                                >
                                  <input
                                    aria-invalid={false}
                                    className="Inputstyles__StyledInput-hc8ttv-2 bUBKOu"
                                    onBlur={[Function]}
                                    onChange={[Function]}
                                    onFocus={[Function]}
                                    onKeyPress={[Function]}
                                    onScroll={[Function]}
                                    size="l"
                                    value="17.06.2020"
                                  />
                                </Inputstyles__StyledInput>
                              </div>
                            </Inputstyles__Box>
                          </NativeInput>
                        </div>
                      </Inputstyles__Wrapped>
                      <Inputstyles__Postfix
                        active={false}
                        size="l"
                      >
                        <div
                          className="Inputstyles__Postfix-hc8ttv-4 dvGfwK"
                          size="l"
                        >
                          <Calendar
                            color="currentColor"
                            cursor="pointer"
                            key=".2"
                            onClick={[Function]}
                            size="m"
                          >
                            <svg
                              cursor="pointer"
                              fill="currentColor"
                              height={20}
                              onClick={[Function]}
                              viewBox="0 0 24 24"
                              width={20}
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <defs />
                              <path
                                d="M10 3v2h4V3h2v2h4a1 1 0 011 1v14a1 1 0 01-1 1H4a1 1 0 01-1-1V6a1 1 0 011-1h4V3h2zm8.999 10h-14L5 19h14l-.001-6zM16 15a1 1 0 110 2 1 1 0 010-2zM8 7H5l-.001 4h14L19 7h-3v2h-2V7h-4v2H8V7z"
                              />
                            </svg>
                          </Calendar>
                        </div>
                      </Inputstyles__Postfix>
                    </label>
                  </Inputstyles__Wrapper>
                </InputWrapper>
              </Input>
            </InputMask>
          </InputDate>
        </div>
      </Datepickerstyles__InputBox>
    </div>
  </Datepickerstyles__Box>
</Datepicker>
`;

exports[`Datepicker /w date not in range and not fix value 1`] = `
<Datepicker
  isCalendarOpen={false}
  maxDate={2020-06-14T21:00:00.000Z}
  minDate={2020-06-01T21:00:00.000Z}
  todayButton={true}
  value={2020-06-16T21:00:00.000Z}
>
  <Datepickerstyles__Box
    onBlur={[Function]}
  >
    <div
      className="Datepickerstyles__Box-sc-15d4dji-0 dajHNb"
      onBlur={[Function]}
    >
      <Datepickerstyles__InputBox
        onClick={[Function]}
      >
        <div
          className="Datepickerstyles__InputBox-sc-15d4dji-1 jpzdav"
          onClick={[Function]}
        >
          <InputDate
            active={false}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyPress={[Function]}
            postfix={
              <Calendar
                color="currentColor"
                cursor="pointer"
                onClick={[Function]}
                size="m"
              />
            }
            value={2020-06-16T21:00:00.000Z}
          >
            <InputMask
              active={false}
              blocks={
                Object {
                  "DATE": Object {
                    "mask": [Function],
                    "max": 9998-12-31T21:00:00.000Z,
                    "min": 1899-12-31T21:29:43.000Z,
                  },
                }
              }
              design="outline"
              format={[Function]}
              lazy={true}
              mask="DATE"
              onChange={[Function]}
              onFocus={[Function]}
              onKeyPress={[Function]}
              parse={[Function]}
              postfix={
                <Calendar
                  color="currentColor"
                  cursor="pointer"
                  onClick={[Function]}
                  size="m"
                />
              }
              size="l"
              smartPlaceholder="17.06.2020"
              value="17.06.2020"
            >
              <Input
                active={false}
                design="outline"
                onChange={[Function]}
                onFocus={[Function]}
                onKeyPress={[Function]}
                onUpdated={[Function]}
                postfix={
                  <Calendar
                    color="currentColor"
                    cursor="pointer"
                    onClick={[Function]}
                    size="m"
                  />
                }
                size="l"
                smartPlaceholder="17.06.2020"
                value="17.06.2020"
              >
                <InputWrapper
                  active={false}
                  design="outline"
                  size="l"
                >
                  <Inputstyles__Wrapper
                    active={false}
                    design="outline"
                    size="l"
                  >
                    <label
                      className="Inputstyles__Wrapper-hc8ttv-6 daAZxl"
                      size="l"
                    >
                      <Inputstyles__Wrapped>
                        <div
                          className="Inputstyles__Wrapped-hc8ttv-5 lhoBGY"
                        >
                          <NativeInput
                            active={false}
                            aria-invalid={false}
                            design="outline"
                            key=".1"
                            onBlur={[Function]}
                            onChange={[Function]}
                            onFocus={[Function]}
                            onKeyPress={[Function]}
                            onUpdated={[Function]}
                            size="l"
                            smartPlaceholder="17.06.2020"
                            value="17.06.2020"
                          >
                            <Inputstyles__Box>
                              <div
                                className="Inputstyles__Box-hc8ttv-1 bzkyqR"
                              >
                                <Inputstyles__PlaceholderWrapper>
                                  <div
                                    className="Inputstyles__PlaceholderWrapper-hc8ttv-7 frQCmf"
                                  >
                                    <Inputstyles__Placeholder
                                      design="outline"
                                      size="l"
                                    >
                                      <div
                                        className="Inputstyles__Placeholder-hc8ttv-0 dDQzQZ"
                                        size="l"
                                      >
                                        <Inputstyles__Invisible>
                                          <span
                                            className="Inputstyles__Invisible-hc8ttv-8 bCdWvB"
                                          >
                                            17.06.2020
                                          </span>
                                        </Inputstyles__Invisible>
                                        <span />
                                      </div>
                                    </Inputstyles__Placeholder>
                                  </div>
                                </Inputstyles__PlaceholderWrapper>
                                <Inputstyles__StyledInput
                                  active={false}
                                  aria-invalid={false}
                                  design="outline"
                                  onBlur={[Function]}
                                  onChange={[Function]}
                                  onFocus={[Function]}
                                  onKeyPress={[Function]}
                                  onScroll={[Function]}
                                  size="l"
                                  value="17.06.2020"
                                >
                                  <input
                                    aria-invalid={false}
                                    className="Inputstyles__StyledInput-hc8ttv-2 bUBKOu"
                                    onBlur={[Function]}
                                    onChange={[Function]}
                                    onFocus={[Function]}
                                    onKeyPress={[Function]}
                                    onScroll={[Function]}
                                    size="l"
                                    value="17.06.2020"
                                  />
                                </Inputstyles__StyledInput>
                              </div>
                            </Inputstyles__Box>
                          </NativeInput>
                        </div>
                      </Inputstyles__Wrapped>
                      <Inputstyles__Postfix
                        active={false}
                        size="l"
                      >
                        <div
                          className="Inputstyles__Postfix-hc8ttv-4 dvGfwK"
                          size="l"
                        >
                          <Calendar
                            color="currentColor"
                            cursor="pointer"
                            key=".2"
                            onClick={[Function]}
                            size="m"
                          >
                            <svg
                              cursor="pointer"
                              fill="currentColor"
                              height={20}
                              onClick={[Function]}
                              viewBox="0 0 24 24"
                              width={20}
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <defs />
                              <path
                                d="M10 3v2h4V3h2v2h4a1 1 0 011 1v14a1 1 0 01-1 1H4a1 1 0 01-1-1V6a1 1 0 011-1h4V3h2zm8.999 10h-14L5 19h14l-.001-6zM16 15a1 1 0 110 2 1 1 0 010-2zM8 7H5l-.001 4h14L19 7h-3v2h-2V7h-4v2H8V7z"
                              />
                            </svg>
                          </Calendar>
                        </div>
                      </Inputstyles__Postfix>
                    </label>
                  </Inputstyles__Wrapper>
                </InputWrapper>
              </Input>
            </InputMask>
          </InputDate>
        </div>
      </Datepickerstyles__InputBox>
    </div>
  </Datepickerstyles__Box>
</Datepicker>
`;

exports[`Datepicker /w invalid 1`] = `
<div
  class="Datepickerstyles__Box-sc-15d4dji-0 dajHNb"
>
  <div
    class="Datepickerstyles__InputBox-sc-15d4dji-1 jpzdav"
  >
    <label
      class="Inputstyles__Wrapper-hc8ttv-6 iaA-dFS"
    >
      <div
        class="Inputstyles__Wrapped-hc8ttv-5 lhoBGY"
      >
        <div
          class="Inputstyles__Box-hc8ttv-1 bzkyqR"
        >
          <div
            class="Inputstyles__PlaceholderWrapper-hc8ttv-7 frQCmf"
          >
            <div
              class="Inputstyles__Placeholder-hc8ttv-0 dDQzQZ"
            >
              ДД.ММ.ГГГГ
            </div>
          </div>
          <input
            aria-invalid="true"
            class="Inputstyles__StyledInput-hc8ttv-2 bUBKOu"
            value=""
          />
        </div>
      </div>
      <div
        class="Inputstyles__Postfix-hc8ttv-4 dvGfwK"
      >
        <svg
          cursor="pointer"
          fill="currentColor"
          height="20"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs />
          <path
            d="M10 3v2h4V3h2v2h4a1 1 0 011 1v14a1 1 0 01-1 1H4a1 1 0 01-1-1V6a1 1 0 011-1h4V3h2zm8.999 10h-14L5 19h14l-.001-6zM16 15a1 1 0 110 2 1 1 0 010-2zM8 7H5l-.001 4h14L19 7h-3v2h-2V7h-4v2H8V7z"
          />
        </svg>
      </div>
    </label>
  </div>
</div>
`;

exports[`Datepicker /w not complete 1`] = `
<div
  class="Datepickerstyles__Box-sc-15d4dji-0 dajHNb"
>
  <div
    class="Datepickerstyles__InputBox-sc-15d4dji-1 jpzdav"
  >
    <label
      class="Inputstyles__Wrapper-hc8ttv-6 daAZxl"
    >
      <div
        class="Inputstyles__Wrapped-hc8ttv-5 lhoBGY"
      >
        <div
          class="Inputstyles__Box-hc8ttv-1 bzkyqR"
        >
          <div
            class="Inputstyles__PlaceholderWrapper-hc8ttv-7 frQCmf"
          >
            <div
              class="Inputstyles__Placeholder-hc8ttv-0 dDQzQZ"
            >
              22.02.ГГГГ
            </div>
          </div>
          <input
            aria-invalid="false"
            class="Inputstyles__StyledInput-hc8ttv-2 bUBKOu"
            value="22.02"
          />
        </div>
      </div>
      <div
        class="Inputstyles__Postfix-hc8ttv-4 dvGfwK"
      >
        <svg
          cursor="pointer"
          fill="currentColor"
          height="20"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs />
          <path
            d="M10 3v2h4V3h2v2h4a1 1 0 011 1v14a1 1 0 01-1 1H4a1 1 0 01-1-1V6a1 1 0 011-1h4V3h2zm8.999 10h-14L5 19h14l-.001-6zM16 15a1 1 0 110 2 1 1 0 010-2zM8 7H5l-.001 4h14L19 7h-3v2h-2V7h-4v2H8V7z"
          />
        </svg>
      </div>
    </label>
  </div>
</div>
`;

exports[`Datepicker 1`] = `
<div
  class="Datepickerstyles__Box-sc-15d4dji-0 dajHNb"
>
  <div
    class="Datepickerstyles__InputBox-sc-15d4dji-1 jpzdav"
  >
    <label
      class="Inputstyles__Wrapper-hc8ttv-6 daAZxl"
    >
      <div
        class="Inputstyles__Wrapped-hc8ttv-5 lhoBGY"
      >
        <div
          class="Inputstyles__Box-hc8ttv-1 bzkyqR"
        >
          <div
            class="Inputstyles__PlaceholderWrapper-hc8ttv-7 frQCmf"
          >
            <div
              class="Inputstyles__Placeholder-hc8ttv-0 dDQzQZ"
            >
              ДД.ММ.ГГГГ
            </div>
          </div>
          <input
            aria-invalid="false"
            class="Inputstyles__StyledInput-hc8ttv-2 bUBKOu"
            value=""
          />
        </div>
      </div>
      <div
        class="Inputstyles__Postfix-hc8ttv-4 dvGfwK"
      >
        <svg
          cursor="pointer"
          fill="currentColor"
          height="20"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs />
          <path
            d="M10 3v2h4V3h2v2h4a1 1 0 011 1v14a1 1 0 01-1 1H4a1 1 0 01-1-1V6a1 1 0 011-1h4V3h2zm8.999 10h-14L5 19h14l-.001-6zM16 15a1 1 0 110 2 1 1 0 010-2zM8 7H5l-.001 4h14L19 7h-3v2h-2V7h-4v2H8V7z"
          />
        </svg>
      </div>
    </label>
  </div>
</div>
`;
